File: h5gl/new/src/pan3d/display/particle/model/Display3DModelPartilce.ts
module Pan3d {
    export class Display3DModelPartilce extends Display3DParticle {
 
        protected _resultUvVec: Array<number>;
        public constructor(value:Scene3D) {
            super(value);
            this._resultUvVec = new Array(2);
        }

        public get modeldata(): ParticleModelData {
            return <ParticleModelData>this.data;
        }

        public creatData(): void {
            this.data = new ParticleModelData(this.scene3D);
        }
        public setVc(): void {

            super.setVc();
            var  ctx:Context3D= this.scene3D.context3D;
            this.setViewCamModeMatr3d();
            this.updateRotaionMatrix();
            ctx.setVcMatrix4fv(this.shader,"rotMatrix",this._rotationMatrix.m);

        }
      
        public updateRotaionMatrix(): void {
            this._rotationMatrix.identity();

            if (this.data._watchEye) {
                this.timeline.inverAxisRotation(this._rotationMatrix);

                this._rotationMatrix.prependRotation(-this.scene3D.cam3D.rotationY, Vector3D.Y_AXIS);

                this._rotationMatrix.prependRotation(-this.scene3D.cam3D.rotationX, Vector3D.X_AXIS);
            }

            if (this.data._isZiZhuan) {
                this.timeline.applySelfRotation(this._rotationMatrix, this.data._ziZhuanAngly);
            }
 

        }

        public setVa(): void {
         
            super.setVa();
            var ctx:Context3D=this.scene3D.context3D;
            ctx.setWriteDepth((<ParticleModelData >this.data)._depthMode == 1)
            var tf: boolean =ctx.pushVa(this.data.objData.vertexBuffer);
            if (!tf) {
                ctx.setVaOffset(0, 3, this.data.objData.stride, 0);
                ctx.setVaOffset(1, 2, this.data.objData.stride, 12);
            }
            this.setMaterialTexture();
            ctx.drawCall(this.data.objData.indexBuffer, this.data.objData.treNum);

            ctx.setWriteDepth(false)
          
        }

        
 
 


    }
}

File: four/app/src/main/java/com/pan3d/display/particle/model/Display3DModelPartilce.java
package com.pan3d.display.particle.model;

import com.pan3d.base.Camera3D;
import com.pan3d.base.ObjData;
import com.pan3d.core.Context3D;
import com.pan3d.display.particle.Display3DParticle;
import com.pan3d.scene.Scene3D;

public class Display3DModelPartilce  extends Display3DParticle {

    public Display3DModelPartilce(Scene3D val) {
        super(val);
    }

    @Override
    public void update() {
        super.update();
    }
    @Override
    public void setVc() {
        super.setVc();
        Context3D ctx= scene3D.context3D;
        this.setViewCamModeMatr3d();
        this.updateRotaionMatrix();
        ctx.setVcMatrix4fv(this.shader3D,"rotMatrix",this.rotationMatrix3D.m);
    }
    @Override
    public void setVa() {
        super.setVa();
        Context3D ctx= scene3D.context3D;
        ObjData objData= this.modelData().objData;
        ctx.setVa(this.shader3D,"v3Position",3,objData.vertexBuffer);
        ctx.setVa(this.shader3D,"v2TexCoord",2,objData.uvBuffer);
        ctx.drawCall(objData.indexBuffer,objData.treNum);
    }
    private void updateRotaionMatrix() {
        Camera3D cam= scene3D.camera3D;
        this.rotationMatrix3D.identity();
        if (this.data._watchEye) {
        }
        if (this.data._isZiZhuan) {
            this.timeline.applySelfRotation(this.rotationMatrix3D,this.data._ziZhuanAngly);
        }
    }
    public ParticleModelData modelData(){
        return (ParticleModelData) this.data;
    }
}
